### PR TITLE
Add meta device with inode in self copy check

### PR DESCRIPTION
**Pull request by xychu**
Combine dev number with inode to identifies a file uniquely,
since the same inode could be used at a different filesystem.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -95,26 +95,26 @@ fn reject_self_copies(sources: &[PathBuf], dest: &Path) -> Result<()> {
...
+        .map(|ancestor| fs::metadata(ancestor).map(|meta| format!("{}/{}", meta.dev().to_string(), meta.ino().to_string())));
 
-    // In contrast, we use `fs::symlink_metadata` for `source_inos` because we copy the symlinks
+    // In contrast, we use `fs::symlink_metadata` for `source_dev_inos` because we copy the symlinks
     // themselves, not the underlying files that they point to.
-    let source_inos = sources
+    let source_dev_inos = sources
         .iter()
-        .map(|source| fs::symlink_metadata(source).map(|meta| meta.ino()))
+        .map(|source| fs::symlink_metadata(source).map(|meta| format!("{}/{}", meta.dev().to_string(), meta.ino().to_string())))
```
Comment (Svetlitski) [inline src/lib.rs:108]:
```suggestion
        .map(|source| fs::symlink_metadata(source).map(unique_id)
```

File: src/lib.rs [hunk 2]
```diff
@@ -95,26 +95,26 @@ fn reject_self_copies(sources: &[PathBuf], dest: &Path) -> Result<()> {
...
         .collect::<Box<_>>();
 
     let mut errors = Vec::new();
 
-    for (ancestor, ino) in dest.ancestors().zip(ancestor_inos) {
-        let ino = ino?;
-        for (source, source_ino) in sources.iter().zip(source_inos.as_ref()) {
-            match source_ino {
-                Ok(source_ino) if *source_ino == ino => errors.push(format!(
+    for (ancestor, dev_ino) in dest.ancestors().zip(ancestor_dev_inos) {
```
Comment (Svetlitski) [inline src/lib.rs:113]:
```suggestion
    for (ancestor, id) in dest.ancestors().zip(ancestor_ids) {
```

Again, change this variable name everywhere it occurs.

File: src/lib.rs [hunk 3]
```diff
@@ -95,26 +95,26 @@ fn reject_self_copies(sources: &[PathBuf], dest: &Path) -> Result<()> {
...
+        let dev_ino = dev_ino?;
+        for (source, source_dev_ino) in sources.iter().zip(source_dev_inos.as_ref()) {
```
Comment (Svetlitski) [inline src/lib.rs:115]:
```suggestion
        for (source, source_id) in sources.iter().zip(source_ids.as_ref()) {
```

Again with the variable name change.

**Review by Svetlitski — CHANGES_REQUESTED** [review]
This is an interesting edge-case that I hadn't considered. Out of curiosity, did you get unlucky enough to actually encounter this while using `fcp`, or is it something that you just noticed while reading through the code?

**Comment by xychu** [comment]
> This is an interesting edge-case that I hadn't considered. Out of curiosity, did you get unlucky enough to actually encounter this while using `fcp`, or is it something that you just noticed while reading through the code?

Got bitten when coping data from USB SSD deivce via  `fcp` :-(
And I'll update my code accordingly.

**Comment by Svetlitski** [comment]
Thank you! I force-pushed just to fix a couple of spelling/phrasing issues in the comments instead of going through the hassle of flagging them in review and asking you to fix them. The failing test-case on CI is just the GitHub actions cache not working, so I'm going to merge this. Thanks for the contribution!

**Pull request merged**